### ISSUE TITLE
 Null key for a Map not allowed in JSON

### ISSUE THREAD

**Issue by eperry** [elastic/logstash#3912]
I am getting 
LogStash::Json::GeneratorError: Null key for a Map not allowed in JSON (use a converting NullKeySerializer?)

It appears this is being called by the redis output but the class that has the problem is JSON.rb which is in the core.

This error is killing logstash instead of erroring and continuing. 

I think I see the problem "environ"=>{nil=>nil,  in the output (I threw in a puts o.to_s to see the object just before the crash

{"file"=>"/proc", "host"=>"hd1pwcs01lx.digital.hbc.com", "type"=>"pidstats", "process"=>{"cmdline"=>"sshd: root@pts/1", "cwd"=>"/", "environ"=>{nil=>nil, "d"=>nil}, "exe"=>"/usr/sbin/sshd", "fd"=>[], "root"=>"/", "nlwp"=>1, "io"=>{"rchar"=>20021385, "wchar"=>20283920, "syscr"=>15614, "syscw"=>15011, "read_bytes"=>0, "write_bytes"=>0, "cancelled_write_bytes"=>0}, "pid"=>16914, "comm"=>"sshd", "state"=>"S", "ppid"=>2360, "pgrp"=>16914, "session"=>16914, "tty_nr"=>0, "tpgid"=>-1, "flags"=>4202752, "minflt"=>1314, "cminflt"=>420, "majflt"=>0, "cmajflt"=>0, "utime"=>80, "stime"=>118, "cutime"=>0, "cstime"=>0, "priority"=>20, "nice"=>0, "itrealvalue"=>0, "starttime"=>202608076, "vsize"=>114155520, "rss"=>1055, "rlim"=>"18446744073709551615", "startcode"=>139991671406592, "endcode"=>139991671959196, "startstack"=>140734550035104, "kstkesp"=>140734550031976, "kstkeip"=>139991625048915, "signal"=>0, "blocked"=>0, "sigignore"=>4096, "sigcatch"=>81926, "wchan"=>"18446744071580571897", "nswap"=>0, "cnswap"=>0, "exit_signal"=>17, "processor"=>17, "rt_priority"=>0, "policy"=>0, "name"=>"sshd", "uid"=>0, "euid"=>0, "gid"=>0, "egid"=>0}, "@version"=>"1", "@timestamp"=>"2015-09-12T00:56:32.392Z", "tags"=>["_grokparsefailure"]}

---
logstash 1.5.4-1
java version "1.7.0_85"
OpenJDK Runtime Environment (rhel-2.6.1.3.el6_7-x86_64 u85-b01)
OpenJDK 64-Bit Server VM (build 24.85-b03, mixed mode)

```
LogStash::Json::GeneratorError: Null key for a Map not allowed in JSON (use a converting NullKeySerializer?)
     jruby_dump at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.4-java/lib/logstash/json.rb:45
     jruby_dump at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.4-java/lib/logstash/json.rb:43
        to_json at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.4-java/lib/logstash/event.rb:158
        receive at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-redis-1.0.0/lib/logstash/outputs/redis.rb:157
         handle at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.4-java/lib/logstash/outputs/base.rb:88
    output_func at (eval):949
   outputworker at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.4-java/lib/logstash/pipeline.rb:244
  start_outputs at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.4-java/lib/logstash/pipeline.rb:166
```

I think some kind of error handling should be placed somewhere to gracefully handle this error which is why I am opening this as a bug.

**Comment by eperry** [comment]
Ok, I was able to eliminate the nil => nil in the logstash-input-proc which created the incompatible hash. but this might pop up with other plugins as people create them

**Comment by guyboertje** [comment]
This will be fixed in JrJackson v0.3.5, (LS v2.x?) This is very unusual.  I am not even sure whether the JSON string representation of a nil key should be an empty string `{"": 42}` or null `{"null": 42}`, I can't find any docs providing clarity.  Ruby Json gem produces an empty string, however the reverse parse operation will not produce the original Hash of `{nil => 42}`, but I think using a `null` will.

**Comment by eperry** [comment]
I don't know the syntax of json well but if you used "null"  can you still have keys with the same name
{ "null"=>"42", "null"=>43 } ?

It might be good enough to throw an error like it does, it is just that this exceptions causes logstash to exit.

**Comment by guyboertje** [comment]
@eperry - the code `{ "null"=>"42", "null"=>43 } ` is not legal Ruby :-) the second key will overwrite the first, but I get your meaning. For any JSON implementation `h1 = Json.parse(Json.generate(h2))`, h1 should be identical to h2.  At this time in LS, the JSON implementation does allow LS to dictate how the nil key should be handled, e.g. use empty string or ignore the nil key and its value.  As the author of JrJackson (the JSON implementation that LS currently uses) I will add this functionality after discussion with the rest of the LS team.

**Comment by guyboertje** [comment]
Actually this is fixed on JrJackson 0.3.4, there is a PR to include it in LS 2.0

**Comment by eperry** [comment]
Awesome thanks glad that we dealt with the original problem and cleaned up the edge type items. Thanks
